app/controllers/users/nbp_wallet_controller.rb

```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+require 'rqrcode'
+
+module Users
+  class NbpWalletController < ApplicationController
+    skip_before_action :require_user!
+    skip_after_action :verify_authorized
+    before_action :set_provider_details
+    before_action :set_connector_api
+
+    def connect
+      if session[:relationship_template_id].present?
+        @connector.reject_all_relationships(session[:relationship_template_id])
+      end
+      template_id, truncated_reference = @connector.create_relationship_template
+      session[:relationship_template_id] = template_id
+      session[:relationship_template_tr] = truncated_reference
+
+      @template_url = "nmshd://tr##{truncated_reference}"
+    rescue Enmeshed::ConnectorError, Faraday::ConnectionFailed => e
+      Sentry.capture_exception(e)
+      redirect_to new_user_registration_path, alert: t('common.errors.generic_try_later')
+    end
+
+    def qr_code
+      send_data RQRCode::QRCode.new("nmshd://tr##{session[:relationship_template_tr]}").as_png(border_modules: 0), type: 'image/png'
+    end
+
+    def relationship_status
+      render json: {status: 'missing relationship_template_id'} and return if session[:relationship_template_id].nil?
+
+      relationships = @connector.relationships(session[:relationship_template_id], 'Pending')
+
+      render json: {status: 'waiting'} and return if relationships.empty?
+
+      render json: {status: 'ready'}
+    rescue Enmeshed::ConnectorError, Faraday::ConnectionFailed => e
+      Sentry.capture_exception(e)
+      redirect_to nbp_wallet_connect_users_path, alert: t('common.errors.generic')
+    end
+
+    def finalize
+      relationships = @connector.relationships(session[:relationship_template_id], 'Pending')
+      abort_and_refresh and return if relationships.size != 1
+
+      accept_and_create_user(relationships.first)
+    rescue Enmeshed::ConnectorError, Faraday::ConnectionFailed => e
+      Sentry.capture_exception(e)
+      abort_and_refresh
+    end
+
+    private
+
+    def accept_and_create_user(relationship) # rubocop:disable Metrics/AbcSize
+      relationship_changes = relationship[:changes]
+      abort_and_refresh if relationship_changes.size != 1
+
+      user_attributes = @connector.parse_userdata_in_rel_change(relationship_changes.first)
+
+      @user = User.new_from_omniauth('nbp', @provider_uid, user_attributes)
+      return abort_and_refresh("Could not create User: #{@user.errors.full_messages.join(', ')}") unless @user.save
+
+      @user.identities << UserIdentity.new(omniauth_provider: 'enmeshed', provider_uid: relationship[:peer])
```

Comment:
I discovered one issue of storing the enmeshed address as `UserIdentity`: When deauthorizing an identity (both, in the view and the `OmniAuthCallbacksController`) we check whether only a single identity is set and no password is configured. For NBP users, however, we have no password but two identities (`enmeshed` + `nbp`). My suggestion: For these occasions, first filter for those identities still configured in the app (with `User.omniauth_providers`) and then perform the check regarding a single identity. This way, it is still generic and also works if we decide to remove some provider for whatever reason.